packages/cli/src/modules/chat-hub/chat-hub.attachment.service.ts

```diff
@@ -0,0 +1,154 @@
+import { Service } from '@n8n/di';
+import { BINARY_ENCODING, type IBinaryData } from 'n8n-workflow';
+import { sanitizeFilename } from '@n8n/utils';
+import { BinaryDataService } from 'n8n-core';
+import { Not, IsNull } from '@n8n/typeorm';
+import { ChatHubMessageRepository } from './chat-message.repository';
+import type { ChatMessageId, ChatSessionId } from '@n8n/api-types';
+import type { ChatHubMessage } from './chat-hub-message.entity';
+import { NotFoundError } from '@/errors/response-errors/not-found.error';
+import { BadRequestError } from '@/errors/response-errors/bad-request.error';
+import type Stream from 'node:stream';
+
+@Service()
+export class ChatHubAttachmentService {
+	private readonly maxTotalSizeBytes = 200 * 1024 * 1024; // 200 MB
+
+	constructor(
+		private readonly binaryDataService: BinaryDataService,
+		private readonly messageRepository: ChatHubMessageRepository,
+	) {}
+
+	/**
+	 * Stores attachments through BinaryDataService.
+	 * This populates the 'id' field for attachments. When external storage is used,
+	 * BinaryDataService replaces base64 data with the storage mode string (e.g., "filesystem-v2").
+	 */
+	async store(
+		sessionId: ChatSessionId,
+		messageId: ChatMessageId,
+		attachments: IBinaryData[],
+	): Promise<IBinaryData[]> {
+		let totalSize = 0;
+		const attachmentsWithBuffer: Array<[IBinaryData, Buffer<ArrayBuffer>]> = [];
+
+		for (const attachment of attachments) {
+			if (attachment.id || !attachment.data) {
+				continue;
+			}
+
+			const buffer = Buffer.from(attachment.data, BINARY_ENCODING);
+			totalSize += buffer.length;
+
+			if (totalSize > this.maxTotalSizeBytes) {
+				const maxSizeMB = Math.floor(this.maxTotalSizeBytes / (1024 * 1024));
+
+				throw new BadRequestError(
+					`Total size of attachments exceeds maximum size of ${maxSizeMB} MB`,
+				);
+			}
+
+			if (attachment.fileName) {
+				attachment.fileName = sanitizeFilename(attachment.fileName);
+			}
+
+			attachmentsWithBuffer.push([attachment, buffer]);
+		}
+
+		return await Promise.all(
+			attachmentsWithBuffer.map(
+				async ([attachment, buffer]) =>
+					await this.binaryDataService.store(
+						{ type: 'chat-hub-message-attachment', sessionId, messageId },
+						buffer,
+						attachment,
+					),
+			),
+		);
```

Comment:
I think it would be nicer to do this within the loop, one by one. Mainly to limit the maximum memory needed for this. With this approach, we need to hold all the files in memory twice (as base64 encoded and as binary buffers). If we do this within the loop, we only need to hold a single file in memory twice